### ISSUE TITLE
[Bug]: Android 10 ADB intercept failing for big responses

### ISSUE THREAD

**Issue by luka-kusulja** [httptoolkit/httptoolkit#669]
### Has this been reported before?

- [X] I have checked for existing reports of this issue

### Repro steps

1. Enable Android ADB interceptor
2. Turn on android app
3. App says connection droped
4. HTTP Toolkit shows no error
5. Turn off intercepting
6. Application works as expected

### How often does this bug happen?

Every time

### The desktop OS you're using

Windows 11 for intercep Android 10 as client

### Details of other apps/devices

Using an rooted android 10 emulated device HTTP intercep closes connection if the response is larger than ~25mb

When http toolkit is disconnected the application works as expected, when using intercept (over ADB) the connection is droped

### Error screenshot

No error on client or interceptor

### Any other info?

Didn't crash, didn't show any error

**Comment by pimterry** [comment]
This isn't enough to reproduce the issue @luka-kusulja unfortunately so I can't really investigate this.

Can you explain exactly what app you're intercepting & the version, which steps to follow in the app to trigger the request, and share a screenshot of the resulting request in HTTP Toolkit? The logs from HTTP Toolkit would also be useful, they're available in Help -> View HTTP Toolkit Logs (note that they reset each launch, so you'll need to reproduce the issue again first).

**Comment by luka-kusulja** [comment]
@pimterry yes, sorry, my mistake, I have reproduced the error and here are the details

**OS:** Windows 11
**IDE:** Android Studio Ladybug 2024.2.1
**AVD:** Pixel Tablet API 29 (Android 10, rooted with magisk)
**HTTP Toolkit**: 1.19.1
**Application:** TiviMate IPTV Player 5.1.5 (installed from google play store, no modifications)
**Interceptor used:** Android Device via ADB

"Test interception" working

1. Open TiviMate
2. Click on "Add playlist"
3. Click on "Xtream Codes"
4. Insert "Server address", "Username", "Password"
5. Click next
6. Application hangs

NOTE: The URL used for "Server address" is http not http**S**

Use the same steps, without HTTP Toolkit, application is working as expected.

I think the problem is that the server response has two "Set-Cookie" headers

![Screenshot 2024-12-05 235353](https://github.com/user-attachments/assets/a5ffd680-16c9-441c-97b6-82fcf8a3b608)

In the following request I can see that only 1 cookie was sent from the client to the server.

While I do understand that having two "Set-Cookie" headers is not normal server behaviour HTTP Toolkit should forward all headers to the client not ommit them.

I will post part of the "last-run.log" file which I got from Help -> View HTTP Toolkit Logs
```
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
Android command [ 'sh', '/data/local/tmp/htk-root-test.sh' ] returned `uid=2000(shell) gid=2000(shell) groups=2000(shell),1004(input),1007(log),1011(adb),1015(sdcard_rw),1028(sdcard_r),3001(net_bt_admin),3002(net_bt),3003(inet),3006(net_bw_stats),3009(readproc),3011(uhid) context=u:r:shell:s0`
Android command [ 'su', '-c', 'sh /data/local/tmp/htk-root-test.sh' ] returned `uid=0(root) gid=0(root) groups=0(root) context=u:r:magisk:s0`
Android command [ 'su', '-c', 'sh', '/data/local/tmp/htk-root-test.sh' ] returned `uid=0(root) gid=0(root) groups=0(root) context=u:r:magisk:s0`
Android command [ 'su', 'root', 'sh', '/data/local/tmp/htk-root-test.sh' ] threw Timeout for ADB command su,root,sh,/data/local/tmp/htk-root-test.sh
Timeout for ADB command su,root,sh,/data/local/tmp/htk-root-test.sh
Android command [ 'su', 'root', 'sh /data/local/tmp/htk-root-test.sh' ] threw Timeout for ADB command su,root,sh /data/local/tmp/htk-root-test.sh
Timeout for ADB command su,root,sh /data/local/tmp/htk-root-test.sh
/system cacert exists but mismatched
Android command [ 'rm', '-f', '/data/local/tmp/htk-root-test.sh' ] returned ``
Android command [ 'ls', '/apex/com.android.conscrypt' ] returned `apex_manifest.json apex_pubkey javalib lib`
No need for /apex cacerts injection
Adding cert file as /data/local/tmp/4f74014f.0
Android command [ 'su', '-c', 'sh /data/local/tmp/htk-inject-system-cert.sh' ] returned `
---
Injecting certificate:
System cacerts setup completed
System cert successfully injected
---`
Cert injected
Android command [ 'su', '-c', 'sh /data/local/tmp/htk-set-chrome-flags.sh' ] returned `Chrome flags script completed`
Chrome flags script completed

Android command [ 'su', '-c', 'am force-stop com.android.chrome' ] returned ``
Android Chrome flags set
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
Android command [ 'su', 'root', 'sh', '/data/local/tmp/htk-root-test.sh' ] returned `generic_x86:/ #`
Android command [ 'su', 'root', 'sh /data/local/tmp/htk-root-test.sh' ] returned `generic_x86:/ #`
emulator-5554:8000->8000 ADB tunnel failed Failure: 'closed'
emulator-5554:8000->8000 ADB tunnel failed Failure: 'closed'
emulator-5554:8000->8000 ADB tunnel failed Failure: 'device 'emulator-5554' not found'
emulator-5554:8000->8000 ADB tunnel failed Failure: 'device 'emulator-5554' not found'
emulator-5554:8000->8000 ADB tunnel failed Failure: 'device 'emulator-5554' not found'
emulator-5554:8000->8000 tunnel disconnected
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.19.1
```

**Comment by pimterry** [comment]
Multiple set-cookies should be fine, that's not uncommon. I've just tested with a tiny local server (just returning a test response with two set-cookie headers) and curl, and that does return both headers exactly as expected when proxying.

I'd be happy to reproduce this directly, but by the sounds of it I can't with the details above without my own IPTV server unfortunately.

I'm a bit confused though, since previously you said it was an issue with large requests, and now we're discussing cookies. From your description, it sounds like it's the request sent with the cookies (the one _after_ the set-cookie response) that's failing unexpectedly, is that right?

Can you share a screenshot of the full details of the failing request, and an exported HAR for it? You can send that by email if there's secrets there (tim @ httptoolkit.com)

It would also be interesting to check whether this fails when resent via the Send page, and when sent by exporting as curl and running directly on the command line (without HTTP Toolkit interception).

**Comment by luka-kusulja** [comment]
Yes you would need access to the IPTV service to properly test this out, I will contact you by mail to sort the login.

Yes the first post was tied to this issue [https://github.com/httptoolkit/httptoolkit/issues/670](https://github.com/httptoolkit/httptoolkit/issues/670) because what I saw was application not working and empty resonse, then by using Postman i saw that the response was 260mb so I incorrectly assumed that HTTP Toolkit can't handle big responses, my mistake sorry about that.

The cookie issue is my current tought, but if you said you have tested it out and its working I don't know why the application doesn't send the second cookie (only the first one) in the next request when intercepting and working when not.

The HAR files and logs contain URLs, usernames / password which I will send you over by email for obvious reasons 😄

Thank you for your time in investigating this.
